### PR TITLE
Fix unreachable DUT mgmt_ip in IPv6 only test case

### PR DESCRIPTION
**Pull request by maipbui**
### Description of PR

Summary: Fix unreachable DUT mgmt_ip in IPv6 only test case
Fixes # (issue)

Microsoft ADO 27674168, 27674176
### Type of change

- [X] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311

#### What is the motivation for this PR?
https://github.com/sonic-net/sonic-mgmt/pull/12303 introduced a bug that DUT mgmt_ipv4 is unreachable in a IPv6 only test case, causing test_rw_user_ipv6_only, test_ro_user_ipv6_only, test_snmp_ipv6_only failed with `paramiko.ssh_exception.NoValidConnectionsError` or `socket.timeout: timed out`
#### How did you do it?
If mgmt_ip is unreachable, try mgmt_ipv6
#### How did you verify/test it?
Verified on internal and 202305 branch on physical DUT

### CODE REVIEW

File: tests/common/utilities.py [hunk 1]
```diff
@@ -467,6 +467,14 @@ def get_test_server_visible_vars(inv_files, server):
     return vm.get_vars(host=test_server_host)
 
 
+def is_ip_reachable(duthost, ip_address):
```
Comment (yejianquan) [inline tests/common/utilities.py:470]: get_dut_current_passwd, the ssh session is from test agent to DUT, in other words sonic-mgmt-container -> DUT.
But the detection here is from DUT -> DUT. 
Which is not 100% aligned with the get_dut_current_passwd logic.

File: tests/conftest.py [hunk 2]
```diff
@@ -764,8 +765,13 @@ def creds_on_dut(duthost):
...
     creds["ansible_altpasswords"] = []
 
+    if is_ip_reachable(duthost, duthost.mgmt_ip):
+        dutip = duthost.mgmt_ip
+    else:
+        dutip = duthost.mgmt_ipv6
+
     passwords = creds["ansible_altpasswords"] + [creds["sonicadmin_password"]]
-    creds['sonicadmin_password'] = get_dut_current_passwd(duthost.mgmt_ip, creds['sonicadmin_user'], passwords)
+    creds['sonicadmin_password'] = get_dut_current_passwd(dutip, creds['sonicadmin_user'], passwords)
```
Comment (yejianquan) [inline tests/conftest.py:774]: I would suggest enhancing get_dut_current_passwd, to receive both IPv4 ip and IPv6 ip and internally retry.
An not that similar example: https://github.com/sonic-net/sonic-mgmt/blob/3be50896eeff945f32d6895fcbbaaf935017e713/tests/common/fixtures/duthost_utils.py#L677

Comment (maipbui) [reply tests/conftest.py]: I updated get_dut_current_passwd, please help re-review

**Review by qiluo-msft — APPROVED** [review]
Please also check with other reviewers.

File: tests/common/utilities.py [hunk 3]
```diff
@@ -1151,20 +1153,24 @@ def _paramiko_ssh(ip_address, username, passwords):
...
     return ssh
 
 
-def get_dut_current_passwd(ip_address, username, passwords):
-    _, pwd = _paramiko_ssh(ip_address, username, passwords)
-    return pwd
+def get_dut_current_passwd(ipv4_address, ipv6_address, username, passwords):
+    try:
+        return _paramiko_ssh(ipv4_address, username, passwords)
+    except (AuthenticationException, NoValidConnectionsError, socket.timeout) as e:
```
Comment (yejianquan) [inline tests/common/utilities.py:1173]: This block catches too many exceptions, which definitely works, but not that clear and efficient enough, suggest(please verify the syntax, I didn't test it)：

```
    except AuthenticationException:
        raise
    else: 
        return _paramiko_ssh(ipv6_address, username, passwords)
```
If we catch Authentication error, it means the connection by IPv6 is good, no need to retry with IPv6

**Comment by sdszhang** [comment]
@maipbui  Can you run the test in KVM setup?  I tested this change locally in KVM t0 testbed, there is a failure :

$./run_tests.sh -n vms-kvm-t0 -d vlab-01 -c ip/test_mgmt_ipv6_only.py::test_ro_user_ipv6_only -f vtestbed.yaml -i ../ansible/veos_vtb -u -E -m individual -u
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

output = {'failed': True, 'changed': True, 'stdout': '', 'stderr': "Warning: Permanently added 'fec0::ffff:afa:1' (RSA) to the ...fec0::ffff:afa:1' (RSA) to the list of known hosts.", 'Permission denied, please try again.'], '_ansible_no_log': None}
exp_val1 = 'test', exp_val2 = 'remote_user'

    def check_output(output, exp_val1, exp_val2):
>       pytest_assert(not output['failed'], output['stderr'])
E       Failed: Warning: Permanently added 'fec0::ffff:afa:1' (RSA) to the list of known hosts.
E       Permission denied, please try again.

exp_val1   = 'test'
exp_val2   = 'remote_user'
output     = {'failed': True, 'changed': True, 'stdout': '', 'stderr': "Warning: Permanently added 'fec0::ffff:afa:1' (RSA) to the ...fec0::ffff:afa:1' (RSA) to the list of known hosts.", 'Permission denied, please try again.'], '_ansible_no_log': None}

File: tests/common/utilities.py [hunk 4]
```diff
@@ -1151,20 +1152,24 @@ def _paramiko_ssh(ip_address, username, passwords):
...
     ssh, pwd = _paramiko_ssh(ip_address, username, passwords)
     return ssh
 
 
-def get_dut_current_passwd(ip_address, username, passwords):
-    _, pwd = _paramiko_ssh(ip_address, username, passwords)
-    return pwd
+def get_dut_current_passwd(ipv4_address, ipv6_address, username, passwords):
+    try:
+        return _paramiko_ssh(ipv4_address, username, passwords)
```
Comment (qiluo-msft) [inline tests/common/utilities.py:1171]: >return [](http://example.com/codeflow?start=8&length=6)

original this function return a password string. you want to return a pair instead? Is it an unexpected behavior change?

Comment (qiluo-msft) [reply tests/common/utilities.py:1171]: If you would like to keep old bahavior, suggest use a different variable name `pwd` -> `passwd`.

Comment (maipbui) [reply tests/common/utilities.py]: updated, thanks for catching this

File: tests/common/utilities.py (hunk 3 above)
Comment (maipbui) [reply tests/common/utilities.py]: Updated, thanks for your suggestion

**Comment by maipbui** [comment]
Tested on 202305 and internal on physical DUT, syncing offline with Dashuai Zhang for kvm test

File: tests/common/utilities.py (hunk 3 above)
Comment (yejianquan) [reply tests/common/utilities.py]: //comment fixed

**Comment by sdszhang** [comment]
> @maipbui I tested this change locally in KVM t0 testbed, there is a failure, can you please take a look? You can follow the steps in #12377 to enable ipv6 in docker for KVM testbed.
> 
> ```
> $./run_tests.sh -n vms-kvm-t0 -d vlab-01 -c ip/test_mgmt_ipv6_only.py::test_ro_user_ipv6_only -f vtestbed.yaml -i ../ansible/veos_vtb -u -E -m individual -u
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> output = {'failed': True, 'changed': True, 'stdout': '', 'stderr': "Warning: Permanently added 'fec0::ffff:afa:1' (RSA) to the ...fec0::ffff:afa:1' (RSA) to the list of known hosts.", 'Permission denied, please try again.'], '_ansible_no_log': None}
> exp_val1 = 'test', exp_val2 = 'remote_user'
> 
>     def check_output(output, exp_val1, exp_val2):
> >       pytest_assert(not output['failed'], output['stderr'])
> E       Failed: Warning: Permanently added 'fec0::ffff:afa:1' (RSA) to the list of known hosts.
> E       Permission denied, please try again.
> 
> exp_val1   = 'test'
> exp_val2   = 'remote_user'
> output     = {'failed': True, 'changed': True, 'stdout': '', 'stderr': "Warning: Permanently added 'fec0::ffff:afa:1' (RSA) to the ...fec0::ffff:afa:1' (RSA) to the list of known hosts.", 'Permission denied, please try again.'], '_ansible_no_log': None}
> ```

You can skip this error. It seems to be separate issues, not related to your change. I will check it further and reach out if needed. Your current code change looks good to me.

**Review by sdszhang — APPROVED** [review]
LGTM

**Review by qiluo-msft — APPROVED** [review]

**Pull request merged**